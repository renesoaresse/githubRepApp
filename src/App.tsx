import React from 'react';
import {SafeAreaView, StatusBar} from 'react-native';
import Routers from './routers';
import {Provider} from 'react-redux';
import {store} from './store/store';

const App = () => (
  <SafeAreaView style={{flex: 1, backgroundColor: 'white'}}>
    <StatusBar barStyle={'light-content'} />
    <Provider store={store}>
      <Routers />
    </Provider>
  </SafeAreaView>
);

export default App;
