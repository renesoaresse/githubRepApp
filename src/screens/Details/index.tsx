import React from 'react';
import {WebView} from 'react-native-webview';
import {RouteProp} from '@react-navigation/native';

type Props = {
  route: RouteProp<{params: {html_url: string}}, 'params'>;
};

export const Details = ({route}: Props) => {
  const {html_url} = route.params;

  return <WebView source={{uri: html_url}} />;
};
