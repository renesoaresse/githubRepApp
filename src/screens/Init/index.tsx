import React, {useState} from 'react';
import {
  TextInput,
  Text,
  Image,
  View,
  FlatList,
  TouchableOpacity,
} from 'react-native';

import {StackNavigationProp} from '@react-navigation/stack';
import {useSelector, useDispatch} from 'react-redux';

import api from '../../services/api';
import {RootState} from '../../store/store';
import {setRepositories} from '../../store/slices/repository';

type Props = {
  navigation: StackNavigationProp<any>;
};

export const Init = ({navigation}: Props) => {
  const [text, setText] = useState('');
  const dispatch = useDispatch();
  const repositories = useSelector(
    (state: RootState) => state.repository.repositories,
  );
  console.log('ðŸš€ ~ file: index.tsx:28 ~ Init ~ repositories:', repositories);

  const searchRepos = async () => {
    const {data} = await api.get(`/search/repositories?q=${text}&per_page=1`);
    dispatch(setRepositories(data.items));
  };

  return (
    <View style={{alignContent: 'center', justifyContent: 'center'}}>
      <TextInput
        value={text}
        onChangeText={setText}
        onBlur={searchRepos}
        onSubmitEditing={searchRepos}
        style={{borderWidth: 1}}
      />
      <FlatList
        data={repositories}
        renderItem={({item}) => (
          <View>
            <TouchableOpacity
              onPress={() =>
                navigation.navigate('DetailsScreen', {
                  html_url: item.owner.html_url,
                })
              }>
              <Image
                source={{uri: item.owner.avatar_url}}
                style={{width: 52, height: 52, borderRadius: 25}}
              />
              <Text>{item.name}</Text>
              <Text>{item.owner.login}</Text>
              <Text>{item.stargazers_count}</Text>
            </TouchableOpacity>
          </View>
        )}
        keyExtractor={item => String(item.id)}
      />
    </View>
  );
};
