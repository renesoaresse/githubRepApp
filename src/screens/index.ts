export {Init} from './Init';
export {Details} from './Details';
