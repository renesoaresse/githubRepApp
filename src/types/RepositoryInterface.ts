export interface RepositoryInterface {
  id: number;
  name: string;
  owner: {
    login: string;
    avatar_url: string;
    html_url: string;
  };
  stargazers_count: number;
}

export interface RepositoryStateInterface {
  repositories: RepositoryInterface[];
}
