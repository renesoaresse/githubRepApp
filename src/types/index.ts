export type {
  RepositoryInterface,
  RepositoryStateInterface,
} from './RepositoryInterface';
