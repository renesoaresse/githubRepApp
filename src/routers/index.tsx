import React from 'react';
import {NavigationContainer} from '@react-navigation/native';
import {createNativeStackNavigator} from '@react-navigation/native-stack';
import {Init, Details} from '../screens';

type StackParamList = {
  Init: undefined;
  Details: {title: string};
};

const Stack = createNativeStackNavigator<StackParamList>();

const Routers = () => (
  <NavigationContainer>
    <Stack.Navigator
      screenOptions={{
        headerStyle: {
          backgroundColor: '#fff',
        },
        headerTintColor: '#000',
        headerTitleStyle: {
          fontWeight: 'bold',
        },
      }}>
      <Stack.Screen name="Init" component={Init} />
      <Stack.Screen
        name="Details"
        component={Details}
        options={({route}) => ({title: route.params.title})}
      />
    </Stack.Navigator>
  </NavigationContainer>
);
export default Routers;
