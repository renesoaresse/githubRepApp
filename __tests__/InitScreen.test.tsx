import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import {Provider} from 'react-redux';
import {configureStore} from '@reduxjs/toolkit';

import {Init} from '../src/screens';
import {RepositoryStateInterface, RepositoryInterface} from '../src/types';

const mockData: RepositoryInterface[] = [
  {
    id: 1,
    name: 'Repo 1',
    stargazers_count: 100,
    owner: {
      login: 'User 1',
      avatar_url: 'https://dummyimage.com/300/09f/fff.png',
      html_url: 'https://dummyimage.com/300/09f/fff.png',
    },
  },
  {
    id: 2,
    name: 'Repo 2',
    stargazers_count: 200,
    owner: {
      login: 'User 2',
      avatar_url: 'https://dummyimage.com/500/09f/fff.png',
      html_url: 'https://dummyimage.com/500/09f/fff.png',
    },
  },
];

jest.mock('react-native-webview', () => ({
  default: () => jest.fn(),
}));

describe('InitScreen', () => {
  let store: RepositoryStateInterface;

  beforeEach(() => {
    store = configureStore({
      reducer: {
        repository: () => ({repositories: mockData}),
      },
    });
  });

  it('renders correctly', () => {
    const tree = render(
      <Provider store={store}>
        <Init navigation={{navigate: jest.fn()}} />
      </Provider>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders a list of repos', () => {
    const {getAllByTestId} = render(
      <Provider store={store}>
        <Init navigation={{navigate: jest.fn()}} />
      </Provider>,
    );

    const repoItems = getAllByTestId('buttom-navigation-repo');
    expect(repoItems).toHaveLength(mockData.length);
  });

  it('navigates to repo details when a repo is selected', () => {
    const navigate = jest.fn();
    const {getAllByTestId} = render(
      <Provider store={store}>
        <Init navigation={{navigate}} />
      </Provider>,
    );

    const firstRepo = getAllByTestId('buttom-navigation-repo')[0];
    fireEvent.press(firstRepo);

    expect(navigate).toHaveBeenCalledWith('Details', {
      html_url: mockData[0].owner.html_url,
      title: mockData[0].name,
    });
  });

  it('renders "Nenhum repositÃ³rios encontrado" when repositories list is empty', () => {
    const noneStore = configureStore({
      reducer: {
        repository: () => ({repositories: []}),
      },
    });

    const {queryByTestId} = render(
      <Provider store={noneStore}>
        <Init navigation={{navigate: jest.fn()}} />
      </Provider>,
    );

    const listEmptyComponent = queryByTestId('none-repo');
    expect(listEmptyComponent).toBeTruthy();
  });
});
