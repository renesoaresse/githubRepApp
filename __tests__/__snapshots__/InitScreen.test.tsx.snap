// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InitScreen renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "white",
      "flex": 1,
      "paddingHorizontal": 16,
      "paddingTop": 16,
    }
  }
>
  <TextInput
    onBlur={[Function]}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Busca por repositÃ³rios"
    style={
      {
        "backgroundColor": "#7676801F",
        "borderRadius": 8,
        "height": 36,
        "paddingHorizontal": 8,
        "paddingVertical": 7,
      }
    }
    value=""
  />
  <RCTScrollView
    ListEmptyComponent={[Function]}
    data={
      [
        {
          "id": 1,
          "name": "Repo 1",
          "owner": {
            "avatar_url": "https://dummyimage.com/300/09f/fff.png",
            "html_url": "https://dummyimage.com/300/09f/fff.png",
            "login": "User 1",
          },
          "stargazers_count": 100,
        },
        {
          "id": 2,
          "name": "Repo 2",
          "owner": {
            "avatar_url": "https://dummyimage.com/500/09f/fff.png",
            "html_url": "https://dummyimage.com/500/09f/fff.png",
            "login": "User 2",
          },
          "stargazers_count": 200,
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    style={
      {
        "marginTop": 25,
      }
    }
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "marginBottom": 17,
              "opacity": 1,
            }
          }
          testID="buttom-navigation-repo"
        >
          <View
            style={
              {
                "alignContent": "space-between",
                "flexDirection": "row",
              }
            }
          >
            <Image
              source={
                {
                  "uri": "https://dummyimage.com/300/09f/fff.png",
                }
              }
              style={
                {
                  "backgroundColor": "#D9D9D9",
                  "borderRadius": 25,
                  "height": 52,
                  "width": 52,
                }
              }
            />
            <View
              style={
                {
                  "alignContent": "center",
                  "borderBottomColor": "#EBEBEB",
                  "borderBottomWidth": 1,
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "marginLeft": 10,
                }
              }
            >
              <View>
                <Text
                  style={
                    {
                      "color": "black",
                      "fontWeight": "600",
                      "marginBottom": 2,
                      "textTransform": "capitalize",
                    }
                  }
                >
                  Repo 1
                </Text>
                <Text
                  style={
                    {
                      "color": "#48484A",
                      "fontWeight": "400",
                      "textTransform": "capitalize",
                    }
                  }
                >
                  User 1
                </Text>
              </View>
              <Text
                style={
                  {
                    "color": "#48484A",
                    "fontWeight": "400",
                  }
                }
              >
                100
                 stars
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "marginBottom": 17,
              "opacity": 1,
            }
          }
          testID="buttom-navigation-repo"
        >
          <View
            style={
              {
                "alignContent": "space-between",
                "flexDirection": "row",
              }
            }
          >
            <Image
              source={
                {
                  "uri": "https://dummyimage.com/500/09f/fff.png",
                }
              }
              style={
                {
                  "backgroundColor": "#D9D9D9",
                  "borderRadius": 25,
                  "height": 52,
                  "width": 52,
                }
              }
            />
            <View
              style={
                {
                  "alignContent": "center",
                  "borderBottomColor": "#EBEBEB",
                  "borderBottomWidth": 1,
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "marginLeft": 10,
                }
              }
            >
              <View>
                <Text
                  style={
                    {
                      "color": "black",
                      "fontWeight": "600",
                      "marginBottom": 2,
                      "textTransform": "capitalize",
                    }
                  }
                >
                  Repo 2
                </Text>
                <Text
                  style={
                    {
                      "color": "#48484A",
                      "fontWeight": "400",
                      "textTransform": "capitalize",
                    }
                  }
                >
                  User 2
                </Text>
              </View>
              <Text
                style={
                  {
                    "color": "#48484A",
                    "fontWeight": "400",
                  }
                }
              >
                200
                 stars
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
